server/common/dao/src/main/java/org/kaaproject/kaa/server/common/dao/service/EventClassServiceImpl.java

```diff
@@ -177,27 +197,34 @@ public int compare(EventSchemaVersionDto o1,
         }
     }
 
-    private void saveEventClassSchema(EventClassFamilyDto eventClassFamilyDto, EventSchemaMetadata eventSchema, int version) {
-        EventClassDto eventClass = new EventClassDto();
+    private void setEventClassProperties(EventClassFamilyDto eventClassFamilyDto, EventClassDto eventClass) {
         eventClass.setTenantId(eventClassFamilyDto.getTenantId());
-        eventClass.setEcfId(eventClassFamilyDto.getId());
-        eventClass.setFqn(eventSchema.getFqn());
-        eventClass.setType(eventSchema.getType());
-        eventClass.setVersion(version);
-        eventClass.setSchema(eventSchema.getSchema());
-        eventClassDao.save(new EventClass(eventClass));
+        eventClass.setVersion(ctlSchemaDao.findById(eventClass.getCtlSchemaId()).getVersion());
+    }
+
+    private void setBackreference(List<EventClassFamilyVersion> ecfvList) {
+        ecfvList.forEach(ecfv -> ecfv.getRecords().forEach(ec -> ec.setEcfv(ecfv)));
     }
 
-    private boolean validateEventClassFamilyFqns(EventClassFamilyDto eventClassFamily, List<String> fqns) {
-        return eventClassDao.validateFqns(eventClassFamily.getTenantId(), eventClassFamily.getId(), fqns);
+    @Override
+    public boolean validateEventClassFamilyFqns(String eventClassFamilyId, List<String> fqns) {
+        List<String> storedFQNs = getFqnListForECF(eventClassFamilyId);
+        for (String storedFQN : storedFQNs) {
+            long duplicatedFqnCount = fqns.stream().filter(fqn -> fqn.equals(storedFQN)).count();
```

Comment:
take a look at https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#findFirst--
Also, usage of Set(HashSet) instead of List for storedFQNs will help to increase performance of the method.